modules/markdown/assets/js/components/markdown-live-preview.jsx

```diff
@@ -0,0 +1,193 @@
+ * External dependencies
+ */
+import endsWith from 'lodash/endsWith';
+import escape from 'lodash/escape';
+import MarkdownIt from 'markdown-it';
+
+/**
+ * Internal dependencies
+ */
+import { saveCaretPosition } from '../utils/caret-management';
+
+const {
+	createElement
+} = window.wp.element;
+
+const markdownIt = new MarkdownIt( 'zero' ).enable( [
+	'heading',
+	'emphasis',
+	'backticks',
+] );
+
+// Redefines the rules applied by the parser to render each token.  This adds
+// to each token's content, the delimiter the user used (which the parser
+// obviously removes in the resulting HTML)
+const setupMarkdownParser = function() {
+	// Adds `_` or `*` to the beginning of the em tag
+	markdownIt.renderer.rules.em_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `<em>${ token.markup }`;
+	};
+	// Adds `_` or `*` to the end of the em tag
+	markdownIt.renderer.rules.em_close = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `${ token.markup }</em>`;
+	};
+	// Adds `__` or `**` to the beginning of the strong tag
+	markdownIt.renderer.rules.strong_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `<strong>${ token.markup }`;
+	};
+	// Adds `__` or `**` to the end of the strong tag
+	markdownIt.renderer.rules.strong_close = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `${ token.markup }</strong>`;
+	};
+	// Wraps inline code tokens with ```
+	markdownIt.renderer.rules.code_inline = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `<code>${ token.markup }${ escape( token.content ) }${ token.markup }</code>`;
+	};
+	// Adds `#`s to the beginning of the heading content
+	markdownIt.renderer.rules.heading_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		const inline_token = tokens[ idx + 1 ];
+		const text_token = inline_token.children[ 0 ];
+		if ( text_token ) {
+			text_token.content = ` ${ text_token.content }`;
+		}
+		return `<${ token.tag }>${ token.markup }`;
+	};
+};
+
+const renderHTML = function( source ) {
+	if ( source ) {
+		return markdownIt.render( source )
+		// The MarkdownIt renderer adds new lines after each paragraph/heading
+		// tag. This affects the restoration of the caret position, so we need
+		// to remove them
+			.split( /\n/ ).join( '' )
+			.trim();
+	}
+};
+
+const stripTrailingNewLines = function( text ) {
+	if ( text ) {
+		return text.replace( /\n{1,2}$/, '' );
+	}
+};
+
+const triggerOnChange = function( evt, source ) {
+	if ( this.props.onChange ) {
+		// Clone event with Object.assign to avoid
+		// "Cannot assign to read only property 'target' of object"
+		evt = Object.assign( {}, evt, {
+			target: {
+				value: source
+			}
+		} );
+		this.props.onChange( evt );
+	}
+};
+
+const sourceIsEmpty = function( source ) {
+	return ! source || '' === source.trim();
+};
+
+const ignoreLastInput = function( source ) {
+	return endsWith( source, ' ' ) || endsWith( source, ' ' );
+};
```

Comment:
Can we represent the space characters differently to make it obvious what they are?

Reply:
Yeah, totally, this is something I had in my list of improvements. I was planning to create some constants using `String.fromCharCode(0x00A0)` to avoid mistakes between the two different space chars.